templates/fine-tune-llm_v2/cookbooks/optimize_cost/README.ipynb

```diff
@@ -0,0 +1,302 @@
+    "\n",
+    "To change these configs you can simply do:\n",
+    "\n",
+    "```\n",
+    "context_length: <ctx_length>\n",
+    "train_batch_size_per_device: <bs>\n",
+    "# You can also change the validation batch size\n",
+    "eval_batch_size_per_device: <bs>\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Deepspeed configs\n",
+    "\n",
+    "By default, we encourage everyone to use Zero-DP stage 3 with parameter and optimizare's state offloading to be the safest in terms of memory consumption. But this setting is not the fastest. You can find other typical deepseepd configurations under `deepspeed_Configs`. A complete doc on all the configurations can be found on [deepspeeds doc page](https://www.deepspeed.ai/docs/config-json/).\n",
+    "\n",
+    "You can try deactivating parameter offloading for smaller models to speed the training up, if you still have room before OOMing or your context length is small enough to leave some room for memory. You can also change the states of Zero-DP to see how they speed the training up for your use-case. In the YAML all you have to do is change something like:\n",
+    "\n",
+    "```\n",
+    "deepspeed:\n",
+    "  config_path: deepspeed_configs/zero_3.json\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Activation checkpointing"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Activation checkpointing allows you decrease memory usage for computing backpropagation by not saving the activations at all the layers and instead recomputing the activations as you are backpropagating. There is some overlaps between backward pass and activation checkpointing to minimize the overhead but at the end of the day you are spending more Flops and it will have some negative impact on throughput.\n",
+    "\n",
+    "So if you have room for GPU memory, one of the ways you can increase throughput is by turning activation checkpoint off. Having said that, most of the time  the impact on memory is so big that we usually see OOMing after turning it off. To turn it off you can update the YAML:\n",
+    "\n",
+    "\n",
+    "```\n",
+    "no_gradient_checkpoint: True\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Gradient Accumulation\n",
+    "\n",
+    "Gradient accumulation is another way of increasing the batch size without using more instances. This can become useful when the increasing batch size directly would require going multi-node and the node-interconnects are slow. It might be better to increase the the gradient accumulation step in these cases to get less slower than going multi-node. To do this you can easily configure the `gradient_accumulation_steps` parameeter in the YAML:\n",
```

Comment:
```suggestion
    "Gradient accumulation is another way of increasing the batch size without using more instances. This can become useful when the increasing batch size directly would require going multi-node and the node-interconnects are slow. It might be better to increase the the gradient accumulation step in these cases to get less slower than going multi-node. To do this you can easily configure the `gradient_accumulation_steps` parameter in the YAML:\n",
```

nit